tests/integration/shutdown.tcl

```diff
@@ -0,0 +1,157 @@
+foreach how {sigterm shutdown} {
+    test "Shutting down master waits for replica to catch up ($how)" {
+        start_server {} {
+            start_server {} {
+                set master [srv -1 client]
+                set master_host [srv -1 host]
+                set master_port [srv -1 port]
+                set master_pid [srv -1 pid]
+                set replica [srv 0 client]
+                set replica_pid [srv 0 pid]
+
+                # Config master.
+                $master config set repl-diskless-sync yes
+                $master config set repl-diskless-sync-delay 1
+                $master config set shutdown-timeout 300 ; # 5min for slow CI
+
+                # Config replica.
+                $replica replicaof $master_host $master_port
+                $replica config set repl-diskless-load swapdb
```

Comment:
you probably want to set this config before calling REPLICAOF, but actually, why do we want this test to insist on diskless replication?
let's let it use the default. (same for the master)

Reply:
OK, I can use the default, but with RDB replication, I need to write much more keys to fill up the socket buffers so that it fails without this PR. The RDB file is more compact.

Reply:
The data loss problem is more serious in diskless setups. (We're using diskless so it's more relevant to us. I'm biased.)

Reply:
i don't understand what difference you're talking about with regards to the socket buffers.
we're not using key-load-delay here, and rely on catching the replica while it's loading the rdb.
we're letting it get a full sync completely, then pause it and add some writes to the master.
we don't care how it does the initial full sync..
am i missing anything?

Reply:
I'm not sure how it works, but when I designed the test case before I started implementing anything and I expected it to fail but it didn't. I turns out that (in unstable) if master shuts down, it flushes the replication buffer and exits. Then, even a few seconds later, the replica can still receive this data. Data loss only occurs if we have more replication data than what fits in the buffer. That's why I added the junk to the replication buffer before the INCR.

Without the diskless config, I needed much more junk before the INCR to make it fail. I don't know why (just guessing).

If we instead just test that the master is reachable (GET or INFO) and check the INFO output as you wrote in another comment, we don't really need the junk and the test case can be faster and more reliable on slow machines, but then we don't really test the same thing. I guess this is enough though.

Reply:
there must be something we're missing, i don't see why diskless should affect that.
the only thing i can think of is that diskless has a re-sync delay (`repl-diskless-sync-delay`), so maybe it dropped and wasn't able to re-sync, in contrast to disk-base that could.

other than that i don't see any reason why the socket buffers will behave differently, but i do think we need to write more than they could swallow.
this shouldn't be slow if you use big keys or pipeline (deferring client), maybe you can use the `proc populate` util.

Reply:
I made a helper function to fill up the OS socket buffers. Please check the assumptions make sense. I'm resolving this for now.

Reply:
LGTM.
i guess we need to run it on daily CI (freebsd, valgrind, etc).
i remember the client eviction PR had some similar issues and in some cases it didn't behave as initially expected.
(maybe we can copy some logic from there, so suffice by running the above tests)